src/systems/hover_interaction.rs

```diff
@@ -0,0 +1,107 @@
+use amethyst::core::cgmath::{MetricSpace, Point3};
+use amethyst::core::GlobalTransform;
+use amethyst::ecs::prelude::{
+    Component, DenseVecStorage, Entities, Entity, Join, Read, ReadStorage, System, Write,
+    WriteStorage,
+};
+use amethyst::renderer::Material;
+use collision::primitive::Primitive3;
+use collision::ContinuousTransformed;
+use systems::MouseRay;
+
+pub struct HoverInteractionSystem;
+
+impl<'a> System<'a> for HoverInteractionSystem {
+    type SystemData = (
+        Entities<'a>,
+        Read<'a, MouseRay>,
+        ReadStorage<'a, GlobalTransform>,
+        WriteStorage<'a, HoverHandler>,
+        WriteStorage<'a, Material>,
+        Write<'a, Option<Hovered>>,
+    );
+
+    fn run(
+        &mut self,
+        (entities, mouse_ray, transforms, mut hover_handlers, mut materials, mut hovered): Self::SystemData,
+    ) {
+        let mut nearest = None;
+        for (entity, hover_handler, transform) in
+            (&*entities, &mut hover_handlers, &transforms).join()
+        {
+            if let Some(point) = hover_handler
+                .bounding_box
+                .intersection_transformed(&mouse_ray.ray, &transform.0)
+            {
+                let collision_distance = mouse_ray.ray.origin.distance2(point);
+                if nearest.map_or(true, |(nearest_distance, _, _)| {
+                    collision_distance < nearest_distance
```

Comment:
btw.: gitgut

Reply:
i see that it's nice to put these in a tuple here, but after you're done finding the nearest one, you don't need the distance anymore, so that part is not that nice anymore. also, you still have two variables, they are still two different things, no matter how you wrap them and whether or not you can assign both in the same line ;)

the fact that you needed a long comment to explain this is to me a even stronger hint that you're overcomplicating this. i also don't entirely understand how mapping the nearest tuple to a boolean with a default value if the tuple does not contain anything is a better expression of your intention than a simple if clause.

perhaps we should talk in person what your intentions were :)